pytorch_lightning/trainer/trainer.py

```diff
@@ -1068,6 +1068,9 @@ def call_setup_hook(self, model: LightningModule) -> None:
             if not called:
                 self.datamodule.setup(stage=state)
 
+        local_rank = self.local_rank if self.world_size > 1 else None
+        self.profiler.setup(local_rank=local_rank, log_dir=self.log_dir)
```

Comment:
the test failures this needs to be called after L1074 since accessing `log_dir` triggers a broadcast and the process groups aren't set up yet? https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/trainer/properties.py#L130-L138

but the accelerator environment [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/trainer/trainer.py#L433
) should call 
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/accelerators/accelerator.py#L72-L78
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/plugins/training_type/ddp.py#L83-L91
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/plugins/training_type/ddp.py#L164-L190
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/plugins/training_type/ddp.py#L234-L242 

is this a race? also triggering a broadcast for each property access can be unexpectedly expensive.

separately, should we should move the datamodule setup to be right before the model setup?

Reply:
Yes, this would be an issue,

A broadcast is needed as only Dummy Logger are created on rank != 0 and they need to be aware of log_dir from rank 0.

```
broadcast for each property access can be unexpectedly expensive
```

@carmocca I think we should introduce a `on_run_stage` hook and change training_type_plugins to call run_stage directly. This will ensure the local_rank is available and we could call `self.profiler.setup` inside.
```
    def on_run_stage(self):
        self.call_hook("on_run_stage")
        self.call_hook(f"on_run_{self._running_stage.value}")

    def run_stage(self):
        results = None
        self.on_run_stage()
        if self.evaluating:
            results = self.run_evaluate()
        elif self.predicting:
            results = self.run_predict()
        else:
            self.run_train()
        return results
```

Reply:
could we set the `log_dir` on the dummy logger whenever they're created without needing the `on_run_stage` ?

Reply:
We still need to `on_run_stage_setup` as local_rank needs to be defined when running with ddp_spawn.
Actually, this hooks is needed to ensure everything is properly setup by all training type plugins.

For `log_dir`, It seems Dummy Logger is used only when `fast_dev_run`. Asking @awaelchli about the broadcast.